### ISSUE TITLE
generate-benchmark-summary fio runs for long time for relatively small dataset 

### ISSUE THREAD

**Issue by ekuric** [distributed-system-analysis/pbench#1016]
```/usr/bin/perl /opt/pbench-agent/bench-scripts/postprocess/generate-benchmark-summary fio``` takes hours ( in this example more than 3h and still running ) to finish. 
dataset ( in /var/lib/pbench-agent/fio_* ) which needs to be processed is only 2.2 GB 

```
# pwd; du -h . |tail -1
/var/lib/pbench-agent/fio_glusterblock-HDD-25GBper-pod-write_rw_2019.01.02T10.49.00
2.2G	.
```

**Comment by portante** [comment]
@ekuric, can you give us a sense of what makes up that 2.2GB?

Is it tool data or result data?

**Comment by ekuric** [comment]
@portante I think that is all what is collected. For example now I have even smaller data set ( pbench collected data ) and it is waiting 3h to finish postprocessing . Once below ``` /usr/bin/perl /opt/pbench-agent/bench-scripts/postprocess/generate-benchmark-summary fio``` is done, test will be finished, but it really takes long. Maybe if we can make this last step for tests to be executed on ```server``` side and that will avoid long wait time and client will be able to run test without waiting on postprocess step. 
``` 
]# pwd; du -h  |tail -2; ps aux | grep pben
/var/lib/pbench-agent/fio_glusterfs-HDD-12GBper-pod-read_bs_4_2019.01.10T05.38.08
694M	./1-read-4KiB
696M	.
root     119457 99.9  0.2 837468 678448 ?       R    07:43 181:47 /usr/bin/perl /opt/pbench-agent/bench-scripts/postprocess/generate-benchmark-summary fio --test-types=read --config=glusterfs-HDD-12GBper-pod-read_bs_4 --clients=10.128.4.5,10.128.4.20,10.130.3.54,10.128.4.14,10.129.5.156,10.129.5.168,10.128.4.10,10.131.3.124,10.129.5.151,10.128.4.7,10.129.5.170,10.128.4.12,10.131.3.140,10.131.3.130,10.131.3.137,10.129.5.163,10.128.4.23,10.130.3.46,10.129.5.157,10.129.5.172,10.131.3.127,10.129.5.158,10.128.4.2,10.130.3.45,10.130.3.55,10.131.3.144,10.131.3.125,10.131.3.143,10.130.3.50,10.129.5.169,10.130.3.53,10.131.3.132,10.130.3.57,10.130.3.33,10.128.4.3,10.129.5.174,10.130.3.47,10.128.4.17,10.131.3.145,10.129.5.175,10.128.5.254,10.131.3.123,10.131.3.136,10.128.4.19,10.128.4.8,10.131.3.122,10.128.4.21,10.130.3.51,10.128.4.4,10.130.3.44,10.129.5.159,10.131.3.139,10.131.3.126,10.131.3.146,10.129.5.167,10.131.3.133,10.129.5.153,10.130.3.34,10.128.4.22,10.130.3.37,10.128.4.15,10.128.5.253,10.129.5.165,10.129.5.171,10.129.5.155,10.129.5.154,10.130.3.38,10.129.5.152,10.130.3.35,10.131.3.131,10.130.3.41,10.129.5.162,10.128.4.6,10.130.3.48,10.128.4.11,10.128.4.18,10.129.5.166,10.128.4.16,10.131.3.128,10.130.3.49,10.131.3.141,10.129.5.164,10.129.5.160,10.130.3.42,10.129.5.173,10.130.3.40,10.130.3.39,10.129.5.161,10.131.3.134,10.130.3.52,10.128.5.252,10.130.3.43,10.131.3.135,10.128.4.9,10.130.3.36,10.131.3.142,10.128.4.13,10.130.3.56,10.131.3.129,10.131.3.138 --targets=/mnt/pvcmount/f2 --ioengine=libaio -b 4 --numjobs=1 -s 12G --pre-iteration-script=./drop-cache.sh /var/lib/pbench-agent/fio_glusterfs-HDD-12GBper-pod-read_bs_4_2019.01.10T05.38.08
root     297750  0.0  0.0 112708   976 pts/0    S+   10:45   0:00 grep --color=auto pben
root     414562  0.0  0.0 113664  2072 ?        S    05:38   0:02 /bin/bash /opt/pbench-agent/bench-scripts/pbench-fio --test-types=read --config=glusterfs-HDD-12GBper-pod-read_bs_4 --clients=10.128.4.5,10.128.4.20,10.130.3.54,10.128.4.14,10.129.5.156,10.129.5.168,10.128.4.10,10.131.3.124,10.129.5.151,10.128.4.7,10.129.5.170,10.128.4.12,10.131.3.140,10.131.3.130,10.131.3.137,10.129.5.163,10.128.4.23,10.130.3.46,10.129.5.157,10.129.5.172,10.131.3.127,10.129.5.158,10.128.4.2,10.130.3.45,10.130.3.55,10.131.3.144,10.131.3.125,10.131.3.143,10.130.3.50,10.129.5.169,10.130.3.53,10.131.3.132,10.130.3.57,10.130.3.33,10.128.4.3,10.129.5.174,10.130.3.47,10.128.4.17,10.131.3.145,10.129.5.175,10.128.5.254,10.131.3.123,10.131.3.136,10.128.4.19,10.128.4.8,10.131.3.122,10.128.4.21,10.130.3.51,10.128.4.4,10.130.3.44,10.129.5.159,10.131.3.139,10.131.3.126,10.131.3.146,10.129.5.167,10.131.3.133,10.129.5.153,10.130.3.34,10.128.4.22,10.130.3.37,10.128.4.15,10.128.5.253,10.129.5.165,10.129.5.171,10.129.5.155,10.129.5.154,10.130.3.38,10.129.5.152,10.130.3.35,10.131.3.131,10.130.3.41,10.129.5.162,10.128.4.6,10.130.3.48,10.128.4.11,10.128.4.18,10.129.5.166,10.128.4.16,10.131.3.128,10.130.3.49,10.131.3.141,10.129.5.164,10.129.5.160,10.130.3.42,10.129.5.173,10.130.3.40,10.130.3.39,10.129.5.161,10.131.3.134,10.130.3.52,10.128.5.252,10.130.3.43,10.131.3.135,10.128.4.9,10.130.3.36,10.131.3.142,10.128.4.13,10.130.3.56,10.131.3.129,10.131.3.138 --targets=/mnt/pvcmount/f2 --ioengine=libaio -b 4 --numjobs=1 -s 12G --pre-iteration-script=./drop-cache.sh
```

**Comment by portante** [comment]
@ekuric, can you dig into `1-read-4KiB` to see what is the biggest file there?

**Comment by ekuric** [comment]
@portante yes, see below. To generate this report took cca 6h, it was in state per previous comment most of that time. 
biggest file is result.json 
you can also check for yourself at 
```
/pbench/public_html/results/b06-h31-1029p/fio-glusterfs-hdd_all_space_occupied_data_per_pod_12G_PVCsize_80GB/fio_glusterfs-HDD-12GBper-pod-read_bs_4_2019.01.10T05.38.08/1-read-4KiB 
```

biggest files 

``` 
167M	./result.json
35M	./sample4/result.json
35M	./sample3/result.json
35M	./sample2/result.json
35M	./sample1/result.json
````

**Comment by portante** [comment]
Great, thanks.

What version of pbench are you using, and can you post those files somewhere we can look at them?

**Comment by ekuric** [comment]
@portante it is 0.54-1g4802e4ba.noarch
```
# rpm -qa |grep pbench
pbench-agent-0.54-1g4802e4ba.noarch
pbench-agent-internal-0.54-1.noarch
pbench-fio-3.12-2.x86_64
pbench-sysstat-11.2.0-1.x86_64

```

**Comment by atheurer** [comment]
I will need to run a perl profiler to see what's going on.  I'll update when I have that information.

**Comment by atheurer** [comment]
This is the problem with our current JSON perl module.  When using JSON-XS module, I can process this data in about 5 seconds instead of the reported 6 hours.  We need to move to JSON-XS, but this does come from EPEL repo, so that may cause some pain for some.  This requirement should only be for the pbench-agents which do post-processing: no "clients" or "servers".

@ekuric if you want to test, you can convert your controller system (where you run the pbench-fio command) in the following way:

yum install yum install perl-JSON-XS

edit /opt/pbench-agent/bench-scripts/postprocess/process-iteration-samples
-change "use JSON;" to "use JSON::XS;"
-change "my $perl_scalar = from_json( $json_text, { utf8  => 1 } );" to "my $perl_scalar = decode_json( $json_text );"
-change "my $json_text   = to_json( \%iteration, { ascii => 1, pretty => 1, canonical => 1 } );" to "my $json_text   = encode_json( \%iteration);"

@k-rister FYI

**Comment by ekuric** [comment]
@atheurer thanks - can confirm it executes faster / better. @bengland2

**Comment by portante** [comment]
Closing in favor of tracking this in #895.
